### DISCUSSION TITLE
How to create a child/proxy class of the client in TypeScript?

### DISCUSSION THREAD

**Discussion by Frederick888** [NodeRedis/node_redis#2841] [Q&A] [answered]
We have a few frequently used methods from the Redis client that we'd like to collection some metrics on, so I'm looking into creating a child/proxy class of the Redis client like
```ts
export class RedisClientWithMetrics extends RedisClientType {
   override async get(key: RedisCommandArgument): Promise<string | null> {
    const res = await super.get(key)
    sendMetrics({ key, hit: !!res })
    return res
  }
}

const client = new RedisClientWithMetrics({/* options */})
client.get('foo')  // has metrics
client.del('foo')  // calls the original del() directly
```

(Pardon my rough pseudo code, quite new to the TypeScript world 🤦)

But now since `RedisClientType` is a type instead of a concrete class, I cannot derive a new class from it. I'm also aware of [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), but that seems quite unwieldy especially considering https://github.com/microsoft/TypeScript/issues/20846.

Can someone please shed some light on me? Maybe I had been looking into the wrong directions and it's actually quite simple? Thank you.

**Comment by leibale** [comment, marked as answer]
As long as you are not changing the function signature you can just override the function without worrying about the types (as they can stay as is)
```javascript
const originalGet = client.get.bind(client);
client.get = async key => {
  const result = await originalGet(key);
  // ...
  return result;
};
```

**Reply by Frederick888** [reply]
Thanks for the response! And looks like I indeed wasn't on the right track!

For the record, since I realised that `get()` was overloaded, I ended up with something like

```ts
const client = createClient({
    url: 'redis://localhost:6379'
})
const originalGet = client.get.bind(client)
const wrappedGet: typeof originalGet = async (...args) => {
    console.log({ args })
    return await originalGet(...args)
}
client.get = wrappedGet
client.connect()

const response = await client.get('foo')

console.log({ response })
client.disconnect()
```
